### ISSUE TITLE
rustfmt without nightly toolchain

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#2128]
When we introduce `rustfmt` we could not get anywhere close to acceptable formatting without using config items that are only available with a nightly toolchain. This puts a burden on devs to have the nightly toolchain installed and brings with it all the hassles / security concerns of doing so - is there a better solution?

- Using stable is not, IMO, a viable solution. The result of formatting with stable is horrible, and there are way too many prickly devs in this repo to cop that.
- Removing `rustfmt` is explicitly out of scope for this issue, that is discussed: https://github.com/rust-bitcoin/rust-bitcoin/pull/2127#pullrequestreview-1681278634

Is there another solution?

**Comment by dpc** [comment]
In Fedimint the Nix dev shell provides a stable toolchain with nightly rustfmt. No one has to do anything.

**Comment by tcharding** [comment]
If Nix guys don't care then is this _really_ a problem or literally just a Guix problem?

**Comment by dpc** [comment]
It's not even all that fancy. If you put `rustfmt` binary from Nightly Rust toolchain into `PATH` before any other `rustfmt`... you get nightly rustfmt.

**Comment by apoelstra** [comment]
@tcharding Nix guys "don't care" because they're willing to pull nightly compiler binaries from Mozilla (and in general, pull binaries from a lot of places). It's a bit glib to say that source reproducability is "literally just a Guix problem".

I'm also very hesitant to use Nix flakes BTW because there are multiple apparently-WONTFIX issues preventing me from using them in my day-to-day life, so if I were to learn them, I'd basically have to learn a whole new dialect of Nix and use both in parallel and keep mental track of which ones are broken in which ways.

**Comment by dpc** [comment]
>  pull nightly compiler binaries from Mozilla 

You can rebuild everything from source with Nix if you want, BTW. And it is being monitored https://nix-community.github.io/trustix/ and reproducibility is a goal as well. And you want to use self-bootstrapped Rust toolchain, you could publish it somewhere as binaries and pull in into Nix and use that instead of ones from Mozilla (I would be interested in helping with it). And with Nix everyone would get that automatically.

> I'm also very hesitant to use Nix flakes BTW because there are multiple apparently-WONTFIX issues preventing me from using them in my day-to-day life, 

What exactly? Other than terrible performance on large repos, I can't think of big blockers.

> I'd basically have to learn a whole new dialect of Nix and use both in parallel and keep mental track of which ones are broken in which ways.

It's not a different dialect of Nix. It's the same Nix, just with some additional structure around it to create something like `Cargo.lock` for Nix, and additional commands to use it. Flakes and non-flakes can interoperate. Lots of projects have flake.nix that lowers down to `default.nix` and `shell.nix` or the other way around, so the amount of extra "stuff" is minimal.

If you have Nix flake, you don't need to force anyone to use it. At Fedimint we do, but it's our own decision.

**Comment by dpc** [comment]
(I don't understand what's the problem with Guix, it should be able to do nightly rustfmt trick just fine)

**Comment by dpc** [comment]
You can't keep pilling up unpragmatic choices and expect things to go well.

I get it, I agree with philosophy and goals all of these choices (self bootstraped compiler, support for older compilers and Debian as most trustworthy distro, and source reproducibility, and security concerns, and so and so), but such is the reality, that you can't go extremist on 20 things at the same time and then expect the world to abandon their goals in favor of yours.

"I want to use a relatively immature programming language, with quickly evolving ecosystem, using a self-bootstrap compiler, that is 2 years behind, but I want it all to develop the code on very niche OSes, and I want it to be easy to setup and accessible and also I want the tooling to do thing exactly how I liked it in my C++ code, but also avoid big formatting-only PRs..."

For some reason it all reminds me of TempleOS. If you go full autarky, you shape your own destiny. :D

**Comment by apoelstra** [comment]
I think it's totally reasonable for people to be able to contribute to this repo without needing a nightly compiler toolchain. This isn't an "unpragmatic choice".

Yes, we encounter a lot of grief due to our MSRV policies which are out of step with the rest of the ecosystem. But normal contributors don't need to think about that (or they rarely do, when CI flags some new language feature that we don't support). It doesn't make sense to conflate this with tooling that is required for all developers.

Rust may be "relatively immature" but it's pretty pathetic for it to be 15 years old without any usable code formatting tool.

**Comment by dpc** [comment]
> Rust may be "relatively immature" but it's pretty pathetic for it to be 15 years old without any usable code formatting tool.

`rustfmt` is very usable. It's almost only in this community where ex-C++-developers sill think that things will explode because `#include`s are in the wrong order, or they are going to overlook a CVE because lines were broken in not their favorite way.

As far as I know C++ tooling had millions years of headstart and still I spent more time dealing with problems involving cross-compiling of rocksdb in Fedimint, then with any Rust problems. And the reason why C++ has so little ecosystem to show for it is because C++ can't ever agree which subset to use, which package manager to use and how to format the code. Generally in my experience pretty much every ecosystem has been catching up with the quality of breath of tooling in Rust, so I don't understand where do you anchor your expectations. :P

Anyway - if you don't want to use nightly rustfmt, then just use a stable one, and accept worse formatting. That would be a pragmatic decision.

**Comment by apoelstra** [comment]
>rustfmt is very usable. It's almost only in this community where ex-C++-developers sill think that things will explode because #includes are in the wrong order, or they are going to overlook a CVE because lines were broken in not their favorite way.

This is neither a fair nor correct characterization of the problems rustfmt brings. rustfmt repeatedly and consistently

* Refomats blocks of hex constants to no longer have any structure
* **moves code comments to other lines of code**
* adds tons of indents to code comments so that the comments are effectively invisible

This is aside from it turning one-word diffs into many-line diffs by agressively reflowing function arguments to collapse or expand them, and doing the same for iterator chains, etc., which would make review harder in any language.

>Generally in my experience pretty much every ecosystem has been catching up with the quality and the breath of tooling in Rust, so I don't understand where do you anchor your expectations. :P

We have a C++-style dependency tree (i.e. basically no dependencies) because the cargo ecosystem is so badly broken that we otherwise could not keep our code compiling for more than six months.

Go has had a formatting tool for a decade which has none of the problems that rustfmt has.

>Anyway - if you don't want to use nightly rustfmt, then just use a stable one, and accept worse formatting. 

We would have better formatting and less pain by just not using rustfmt at all.

**Comment by tcharding** [comment]
How about this solution:

- We disable `rustfmt` in CI
- We explicitly do not comment on code style in review unless it makes the code less readable
- Once every so often I run the formater then manually commit just the bits that are better ("better" defined by review consensus)
- We have a one-ack-for-formatting changes carve-out so only @apoelstra has to review the formatting PRs I do and they can go in quickly
- We use `rustfmt.toml` to disable it so we work with editors
- We have a loose code style policy, for example "comments should use approx 100 column width"

**Comment by junderw** [comment]
> How about this solution:
> 
> * We disable `rustfmt` in CI
> * We explicitly do not comment on code style in review unless it makes the code less readable
> * Once every so often I run the formater then manually commit just the bits that are better ("better" defined by review consensus). (And cry a little bit each time I waste my life doing this.)
> * We have a one-ack-for-formatting changes carve-out so only @apoelstra has to review the formatting PRs I do and they can go in quickly
> * We use `rustfmt.toml` to disable it so we work with editors
> * We have a loose code style policy, for example "comments should use approx 100 column width"

I agree with all but the last one. It clashes with 2.

**Comment by tcharding** [comment]
> I agree with all but the last one. It clashes with 2.

Oh I mean just so that I know what the policy is so I don't have to think about it while doing the formatting PRs. I am more or less of the camp "I don't care what it is as long as its uniform" - seems to reduce the cognitive load I carry while coding.

**Comment by junderw** [comment]
This proposal also places a lot of work on @apoelstra and @tcharding.

As well as any coders who don't touch formatting while coding and just run rustfmt at the end (like myself).

There will be the extra cognitive load of "ok, do I line break here? or not?" "Would it be considered 'unless it makes the code less readable' here or not?"

It seems like these loose rules are explicitly loose in order to reduce cognitive load, but the jump from 0 (leaving it to rustfmt config) to non-zero (even if loose) is a big one...

That said, I think it might be the only way to move forward if we want to remove nightly requirement but also don't want to use stable rustfmt.

**Comment by tcharding** [comment]
> As well as any coders who don't touch formatting while coding and just run rustfmt at the end (like myself).

Yes, I have my editor format on save, I'm pretty sure I have come to rely on that. But I can unlearn it because I only work on rust-bitcoin crates but its a massive PITA for people who work on other crates that use the formatter. This is a step backwards for contributors.

**Comment by dpc** [comment]
> * Refomats blocks of hex constants to no longer have any structure

Skip formatting in places where it makes sense. Job done.

> * **moves code comments to other lines of code**

I've only see it happens when putting comments in weird places (like at the right side of the code, like people do in C & C++ ... - there's a theme here), because the core competency of a system developer is ASCII-art and fighting vertical space inflation.

> * adds tons of indents to code comments so that the comments are effectively invisible

No idea about this one. We have tons of comments, noone complains. Could be stable does something much worse then nightly here?


I think you should disable `rustfmt` and over time move towards familiar and well established ways of doing things like nitpicking on whitespaces and arguing about each other's ASCII-art skills, which is know to be an efficient way to produce high amounts of secure software.

**Comment by realeinherjar** [comment]
> You can rebuild everything from source with Nix if you want, BTW. And it is being monitored https://nix-community.github.io/trustix/ and reproducibility is a goal as well. And you want to use self-bootstrapped Rust toolchain, you could publish it somewhere as binaries and pull in into Nix and use that instead of ones from Mozilla (I would be interested in helping with it). And with Nix everyone would get that automatically.


I am in the middle of a proposal to migrate CI and Dev shells to Nix (https://github.com/bitcoindevkit/bdk/pull/1165)

**Comment by yancyribbens** [comment]
> We disable rustfmt in CI

Seems like a good idea.  Currently most commits are done by developers that just use rust fmt nightly, and so over time it won't make much difference except for maybe some one off commits, and there isn't really any danger in that.  Also add a suggestion to the style policy to use rust fmt nightly.

**Comment by apoelstra** [comment]
> Seems like a good idea. Currently most commits are done by developers that just use rust fmt nightly

This is a little bit tempting. Or we could weaken CI to say "if the rustfmt diff is less than 20 lines it doesn't matter" on the assumption that me or Tobin will address it in another PR. But the premise here is that non-rustfmt'ers won't mess up the formatting so badly that fixing it will wreck others' diffs.

And so no matter what, the rule would be something like "if you don't run rustfmt we'll probably let it go but not if you change too much code" and I think that'll just be unpredictable and annoying. Plus it still will cause unrelated diffs to have formatting changes. Which will be annoying even if it's "not much".

@dpc 
> I've only see it happens when putting comments in weird places (like at the right side of the code, like people do in C & C++

"You're holding it wrong"

> ... - there's a theme here), because the core competency of a system developer is ASCII-art and fighting vertical space inflation.

Because we don't use Orwellian newspeak languages where it's impossible to even express ASCII art without the formatter erasing it.

(Sorry for adding more snark but I was kinda proud of those two)

**Comment by dpc** [comment]
> Because we don't use Orwellian newspeak languages where it's impossible to even express ASCII art without the formatter erasing it.
> 
> (Sorry for adding more snark but I was kinda proud of those two)

Comrade, you need to understand and accept that in building our glorious RiiR utopia, we need Rust developers of the world to work united, keyboard by keyboard, leaving individual differences aside. Sacrifices need to be made.

Unlike the reactionaries, we don't let pointers to arbitrary memeory because of individual preferences. Same way, we don't let an arbitrary formatting. Both are to serve the common interest of the Rust people, not individuals.

These frivolous displays of individualistic self-expression are counterproductive to achieving the collective goal of eradicating memory unsafety.

Council of People's Commissariat for Formatting has provided us with good enough rules and tools, and I'm sure the current Five Year Plan will improve things even further. Just be patient.

Glory to the Crab.

**Comment by junderw** [comment]
```
/*
TODO: This line of code needs to be removed, but until then... here's a stray cat in a box of Japanese tangerines.

 ____ ∧ ∧
|＼ /(´～`)＼< meow
|　|￣￣￣￣￣|
|　|＝みかん＝|
 ＼|＿＿＿＿＿|
*/
```

**Comment by tcharding** [comment]
> on the assumption that me or Tobin will address it in another PR

I'm totally fine to remove `rustfmt` from CI. We can write a script that runs daily that pulls down master, runs `rustfmt` and creates a PR. It solves the review problem too because reviewers can ignore formatting issues knowing the code will be formatted the day after merge. We would need a one-ack carve out if the script makes a PR with my or Andrew's account or else we need a bot so both of us can ack it, I've not been all that interested in learning how to write github actions thus far but perhaps this would be the catalyst.

**Comment by apoelstra** [comment]
concept ACK from me. The bot should have its own account. But I don't really want to do the work :).

Maybe we could get bors to do this? I have a very longstanding TODO to setup bors here..

**Comment by realeinherjar** [comment]
> I'm totally fine to remove `rustfmt` from CI. We can write a script that runs daily that pulls down master, runs `rustfmt` and creates a PR.

This can be a scheduled GH-actions.
I can do a PR in the next week if that is the way you would want to go.

**Comment by apoelstra** [comment]
ACK from me. @tcharding what do you think?

**Comment by apoelstra** [comment]
I think we should start by doing it every week and see how big the diff is.

**Comment by tcharding** [comment]
> This can be a scheduled GH-actions.
> I can do a PR in the next week if that is the way you would want to go.

Legend!

> I think we should start by doing it every week and see how big the diff is.

Every week should be fine I'd say.

**Comment by yancyribbens** [comment]
> We can write a script that runs daily that pulls down master, runs rustfmt and creates a PR.

The downside to this is that it's going to make Git blame less useful.  Trying to figure out what PR or reason a change happened, your going to get details about a format commit instead of details about why the author made the change.

**Comment by yancyribbens** [comment]
Maybe it's possible to rebase the commit to retain the original intent when it's formatted.  Personally, I feel like the formatting should just be decided by who does the most proof of work :P.  It's ok to not have uniform formatting which is already the case with some the skip macros..

**Comment by yancyribbens** [comment]
Is re-writing commit history and requiring a format newspeak?  #philosoraptor

**Comment by apoelstra** [comment]
> Is re-writing commit history and requiring a format newspeak? #1802 

Lol :).

Non-philosophically, it'll erase GPG signatures so we can't do it. And it would undoubtedly cause chaos when people tried to sync their local repo to Github.

> The downside to this is that it's going to make Git blame less useful.

True, but if people just "copy the existing style" hopefully this won't happen too much. git blame is already pretty-hard to use in this repo because of how much module-moving we've done in the last 6 months.
